File: backend/pkg/api/project/handlers.go
package project

import (
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	"backend/pkg/core/project"
	"backend/pkg/database"
	"backend/pkg/database/repository"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// getService creates a project service with repository injection
func getService() *project.Service {
	db := database.DB()
	projectRepo := repository.NewProjectRepository(db)
	tagRepo := repository.NewTagRepository(db)
	return project.NewService(projectRepo, tagRepo)
}

// ListProjects handles GET /projects
// @Summary List projects
// @Description Get a paginated list of projects
// @Tags projects
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param perPage query int false "Items per page" default(20)
// @Success 200 {object} response.SuccessResponse{data=dto.ProjectListResponse}
// @Failure 500 {object} response.SuccessResponse
// @Router /projects [get]
func ListProjects(c *fiber.Ctx) error {
	page := c.QueryInt("page", 1)
	perPage := c.QueryInt("perPage", 20)

	svc := getService()
	result, err := svc.List(c.Context(), page, perPage)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"projects":     result.Projects,
		"total":        result.Total,
		"current_page": result.Page,
		"per_page":     result.PerPage,
	})
}

// GetProject handles GET /projects/:id
// @Summary Get project
// @Description Get a project by ID
// @Tags projects
// @Accept json
// @Produce json
// @Param id path string true "Project ID"
// @Success 200 {object} response.SuccessResponse{data=dto.ProjectResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Router /projects/{id} [get]
func GetProject(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid project ID"))
	}

	svc := getService()
	detail, err := svc.GetDetail(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, detail)
}

// CreateProject handles POST /projects
// @Summary Create project
// @Description Create a new project
// @Tags projects
// @Accept json
// @Produce json
// @Param request body dto.CreateProjectRequest true "Project details"
// @Success 201 {object} response.SuccessResponse{data=dto.ProjectResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /projects [post]
func CreateProject(c *fiber.Ctx) error {
	var req project.CreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	svc := getService()
	result, err := svc.Create(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Created(c, result)
}

// UpdateProject handles PUT /projects/:id
// @Summary Update project
// @Description Update an existing project
// @Tags projects
// @Accept json
// @Produce json
// @Param id path string true "Project ID"
// @Param request body dto.UpdateProjectRequest true "Project update details"
// @Success 200 {object} response.SuccessResponse{data=dto.ProjectResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /projects/{id} [put]
func UpdateProject(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid project ID"))
	}

	var req project.UpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	svc := getService()
	result, err := svc.Update(c.Context(), id, req)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, result)
}

// DeleteProject handles DELETE /projects/:id
// @Summary Delete project
// @Description Delete a project by ID
// @Tags projects
// @Accept json
// @Produce json
// @Param id path string true "Project ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /projects/{id} [delete]
func DeleteProject(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid project ID"))
	}

	// Verify user is authenticated (already done by middleware, but we can get userID if needed)
	_, err = middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, core.UnauthorizedError("Not authenticated"))
	}

	svc := getService()
	if err := svc.Delete(c.Context(), id); err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{"success": true})
}

File: backend/src/api/project/handlers.rs
use axum::{
    Json,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::api::response::SuccessResponse;

use super::{
    dto::{
        ProjectCreateRequest, ProjectDetailResponse, ProjectListQuery, ProjectListResponse,
        ProjectResponse, ProjectUpdateRequest, SuccessFlag,
    },
    error::{ProjectApiError, ProjectAuthenticated},
    state::ProjectState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, ProjectApiError>;

fn project_id(value: &str) -> Result<Uuid, ProjectApiError> {
    Uuid::parse_str(value)
        .map_err(|_| crate::error::AppError::InvalidInput("Invalid project ID".to_owned()).into())
}

#[utoipa::path(
    get,
    path = "/projects",
    params(ProjectListQuery),
    responses(
        (status = 200, body = SuccessResponse<ProjectListResponse>),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    tag = "projects",
    operation_id = "listProjects"
)]
pub async fn list_projects(
    State(state): State<ProjectState>,
    Query(query): Query<ProjectListQuery>,
) -> ApiResult<ProjectListResponse> {
    let (page, per_page) = query.values();
    Ok(Json(SuccessResponse::new(
        state.service().list(page, per_page).await?.into(),
    )))
}

#[utoipa::path(
    get,
    path = "/projects/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<ProjectDetailResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    tag = "projects",
    operation_id = "getProject"
)]
pub async fn get_project(
    State(state): State<ProjectState>,
    Path(id): Path<String>,
) -> ApiResult<ProjectDetailResponse> {
    Ok(Json(SuccessResponse::new(
        state.service().get_detail(project_id(&id)?).await?.into(),
    )))
}

#[utoipa::path(
    post,
    path = "/projects",
    request_body = ProjectCreateRequest,
    responses(
        (status = 201, body = SuccessResponse<ProjectResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "projects",
    operation_id = "createProject"
)]
pub async fn create_project(
    _authenticated: ProjectAuthenticated,
    State(state): State<ProjectState>,
    body: Result<Json<ProjectCreateRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<ProjectResponse>>), ProjectApiError> {
    let Json(request) = body.map_err(|_| ProjectApiError::invalid_body())?;
    let project = state.service().create(request.validate()?).await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(project.into())),
    ))
}

#[utoipa::path(
    put,
    path = "/projects/{id}",
    params(("id" = Uuid, Path)),
    request_body = ProjectUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<ProjectResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "projects",
    operation_id = "updateProject"
)]
pub async fn update_project(
    _authenticated: ProjectAuthenticated,
    State(state): State<ProjectState>,
    Path(id): Path<String>,
    body: Result<Json<ProjectUpdateRequest>, JsonRejection>,
) -> ApiResult<ProjectResponse> {
    let id = project_id(&id)?;
    let Json(request) = body.map_err(|_| ProjectApiError::invalid_body())?;
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .update(id, request.validate()?)
            .await?
            .into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/projects/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "projects",
    operation_id = "deleteProject"
)]
pub async fn delete_project(
    _authenticated: ProjectAuthenticated,
    State(state): State<ProjectState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    state.service().delete(project_id(&id)?).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}
